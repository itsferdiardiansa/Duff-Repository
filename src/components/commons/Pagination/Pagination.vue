<template>
  <div :class="`${prefixClass}-pagination`" v-if="getPagination.isApplicable">
    <div :class="`${prefixClass}-pagination--info`">
      <p>
        Showing
        <span v-text="getPagination.info.from"></span>
        to
        <span v-text="getPagination.info.to"></span>
        of
        <span v-text="getPagination.info.of"></span>
        results
      </p>
    </div>

    <div :class="`${prefixClass}-pagination--page`">
      <span
        class="nav-btn prev"
        v-if="getPagination.prevBtn"
        @click="handlePageChange('prev')"
      >
        <img svg-inline src="@icon/arrow.svg" />
      </span>

      <template v-for="(page, key) in getPagination.pages" :key="key">
        <span
          class="page-number"
          :class="getCurrentPageClass(page)"
          @click="handlePageChange(page)"
          v-text="page"
        ></span>
      </template>

      <span
        class="nav-btn next"
        v-if="getPagination.nextBtn"
        @click="handlePageChange('next')"
      >
        <img svg-inline src="@icon/arrow.svg" />
      </span>
    </div>
  </div>
</template>
<script>
import { computed } from 'vue';
import usePagination from './hooks/usePagination';

export default {
  name: 'CMPagination',
  emits: ['changePage'],
  props: {
    current_page: {
      type: Number,
    },
    total_page: {
      type: Number,
    },
    per_page: {
      type: Number,
    },
  },
  setup(props, { emit }) {
    const getProps = computed(() => ({ ...props }));
    const getCurrentPageClass = page => {
      const { current_page } = props;

      return { active: page === current_page };
    };

    const { getPagination } = usePagination(getProps);

    const handlePageChange = page => {
      const { current_page, per_page } = props;
      let value = { page: null, limit: per_page };

      switch (page) {
        case 'prev':
          value.page = current_page - 1;
          break;
        case 'next':
          value.page = current_page + 1;
          break;
        default:
          value.page = page;
          break;
      }

      emit('changePage', value);
    };

    return {
      getPagination,
      getCurrentPageClass,
      handlePageChange,
    };
  },
};
</script>
<style lang="scss" scoped>
.#{$prefixClass}-pagination {
  @apply px-4 py-3 flex items-center justify-between border-t border-gray-200 sm:px-6;

  &--info {
    p {
      @apply text-sm text-gray-500;

      span {
        @apply font-medium;
      }
    }
  }

  &--page {
    @apply relative z-0 inline-flex shadow-sm -space-x-px;

    .page-number {
      @apply w-10 h-10 relative inline-flex cursor-pointer justify-center items-center border border-gray-300 bg-white text-sm font-medium text-gray-500 hover:bg-gray-50;

      &.active {
        @apply bg-primary border-primary text-gray-200 cursor-default;
      }
    }

    .nav-btn {
      @apply w-10 h-10 relative inline-flex cursor-pointer justify-center items-center;
      @apply mx-0 bg-white text-sm font-medium border border-gray-300 text-gray-500 hover:bg-gray-50;

      &.prev {
        @apply rounded-l-md;

        svg {
          @apply transform rotate-180;
        }
      }

      &.next {
        @apply rounded-r-md;
      }
      svg {
        @apply w-3;
      }
    }
  }
}
</style>
