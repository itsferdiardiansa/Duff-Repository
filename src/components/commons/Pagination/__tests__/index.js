import { mount } from '@vue/test-utils';
import Pagination from '@common/Pagination';

describe('common/Pagination', () => {
  it('will render and match snapshot', () => {
    const props = {
      current_page: 1,
      total_page: 2,
      per_page: 10,
    };

    const Component = mount(<Pagination {...props} />);

    expect(Component.element).toMatchSnapshot();
  });

  it('will not render if it does not have a total_page value', () => {
    const props = {
      current_page: 1,
      total_page: 0,
      per_page: 10,
    };

    const Component = mount(<Pagination {...props} />);

    expect(Component.element).toMatchSnapshot();
  });

  it('it does not render the prev navigation when it is on the first page', async () => {
    const props = {
      current_page: 1,
      total_page: 10,
      per_page: 10,
    };

    const Component = mount(<Pagination {...props} />);

    expect(Component.find('.nav-btn.prev').exists()).toBeFalsy();
    expect(Component.element).toMatchSnapshot();
  });

  it('it does not render the next navigation when it is on the last page', async () => {
    const props = {
      current_page: 10,
      total_page: 10,
      per_page: 10,
    };

    const Component = mount(<Pagination {...props} />);

    expect(Component.find('.nav-btn.next').exists()).toBeFalsy();
    expect(Component.element).toMatchSnapshot();
  });

  it('will return the value when triggered by a click', async () => {
    const handleChange = jest.fn();
    const props = {
      current_page: 3,
      total_page: 10,
      per_page: 10,
    };

    const Component = mount(
      <Pagination {...props} onChangePage={handleChange} />
    );
    const pages = Component.findAll('.page-number');
    const prevBtn = Component.find('.nav-btn.prev');
    const nextBtn = Component.find('.nav-btn.next');

    // Click at page 2
    await pages[1].trigger('click');
    expect(handleChange).toHaveBeenCalledWith({ page: 2, limit: 10 });
    await Component.setProps({ current_page: 2 });

    // Click at page 5
    await prevBtn.trigger('click');
    expect(handleChange).toHaveBeenCalledWith({ page: 1, limit: 10 });
    await Component.setProps({ current_page: 1 });

    // Click at page 5
    await nextBtn.trigger('click');
    expect(handleChange).toHaveBeenCalledWith({ page: 1, limit: 10 });
  });
});
