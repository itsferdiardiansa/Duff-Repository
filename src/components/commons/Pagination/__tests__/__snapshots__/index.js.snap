// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Pagination it does not render the next navigation when it is on the last page 1`] = `
<div
  class="ss-pagination"
>
  <div
    class="ss-pagination--info"
  >
    <p>
       Showing 
      <span>
        91
      </span>
       to 
      <span>
        100
      </span>
       of 
      <span>
        100
      </span>
       results 
    </p>
  </div>
  <div
    class="ss-pagination--page"
  >
    <span
      class="nav-btn prev"
    >
      <img
        svg-inline=""
      />
    </span>
    
    <span
      class="page-number"
    >
      6
    </span>
    <span
      class="page-number"
    >
      7
    </span>
    <span
      class="page-number"
    >
      8
    </span>
    <span
      class="page-number"
    >
      9
    </span>
    <span
      class="page-number active"
    >
      10
    </span>
    
    <!--v-if-->
  </div>
</div>
`;

exports[`common/Pagination it does not render the prev navigation when it is on the first page 1`] = `
<div
  class="ss-pagination"
>
  <div
    class="ss-pagination--info"
  >
    <p>
       Showing 
      <span>
        1
      </span>
       to 
      <span>
        10
      </span>
       of 
      <span>
        100
      </span>
       results 
    </p>
  </div>
  <div
    class="ss-pagination--page"
  >
    <!--v-if-->
    
    <span
      class="page-number active"
    >
      1
    </span>
    <span
      class="page-number"
    >
      2
    </span>
    <span
      class="page-number"
    >
      3
    </span>
    <span
      class="page-number"
    >
      4
    </span>
    <span
      class="page-number"
    >
      5
    </span>
    
    <span
      class="nav-btn next"
    >
      <img
        svg-inline=""
      />
    </span>
  </div>
</div>
`;

exports[`common/Pagination will not render if it does not have a total_page value 1`] = `<!--v-if-->`;

exports[`common/Pagination will render and match snapshot 1`] = `
<div
  class="ss-pagination"
>
  <div
    class="ss-pagination--info"
  >
    <p>
       Showing 
      <span>
        1
      </span>
       to 
      <span>
        10
      </span>
       of 
      <span>
        20
      </span>
       results 
    </p>
  </div>
  <div
    class="ss-pagination--page"
  >
    <!--v-if-->
    
    <span
      class="page-number active"
    >
      1
    </span>
    <span
      class="page-number"
    >
      2
    </span>
    
    <span
      class="nav-btn next"
    >
      <img
        svg-inline=""
      />
    </span>
  </div>
</div>
`;
